# Clusterização

[**Premier League:** Premier League Match Data as well as Financial Data of all clubs in the 18-19](https://www.kaggle.com/thesiff/premierleague1819)


Execute  os  algoritmosde clusterizacao citados a seguir e compare os resultados. 
- Execute os algoritmos de agrupamento K-means e Hierarquico com os seguintes valores de K:2, 5, 10 e 100. Compare os agrupamento resultantes dos 2 algoritmos. 
- Escolha  um  numero  fixo  de  K  e  altere  o  parametro  do  K-Means  referente  ao  numeromaximo  de  iteracoes:  1,  10  e  100  e  o  parametro  de  Linkage  do  Hierarquico,  quaisdiferencas puderam ser observadas? 
- Faca uma comparacao entre os 2 algoritmos, qual voce acha que teve o melhordesempenho e por que? 

**Bonus**:  normalizar  os  dados  e  executar  novamente  os  algoritmos  para  analisar  os  novos resultados gerados. Em seguida, fazer os graficos dos agrupamentos dos dados brutos vs dados normalizados, e um boxplot para mostrar a dispersao destes dados

In [1]:
import pandas as pd
import sklearn as sk
from sklearn import model_selection
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import seaborn as sns


from sklearn.cluster import AgglomerativeClustering
from sklearn.cluster import KMeans

In [2]:
epl_df = pd.read_csv("epl_1819.csv")
transfer_df = pd.read_csv("transfer.csv")

In [3]:
epl_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 44 columns):
Team                             20 non-null object
category                         20 non-null object
general_league_position          20 non-null int64
finance _live_games_televised    20 non-null int64
finance _tv_revenue              20 non-null int64
general_matches_played           20 non-null int64
general_won                      20 non-null int64
general_draw                     20 non-null int64
general_lost                     20 non-null int64
attack_scored                    20 non-null int64
defence_goals_conceeded          20 non-null int64
general_goal_difference          20 non-null int64
general_points                   20 non-null int64
general_squad_size               20 non-null int64
general_squad_average_age        20 non-null float64
general_squad_foreigners         20 non-null int64
finance _team_market             20 non-null int64
finance _market_average   

In [4]:
transfer_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 12 columns):
team        20 non-null object
end_2019    20 non-null float64
end_2018    20 non-null float64
end_2017    20 non-null float64
end_2016    20 non-null float64
end_2015    20 non-null float64
end_2014    20 non-null float64
end_2013    20 non-null float64
end_2012    20 non-null float64
end_2011    20 non-null float64
end_2010    20 non-null float64
Total       20 non-null float64
dtypes: float64(11), object(1)
memory usage: 2.0+ KB


In [5]:
epl_df.head()

Team                        category  general_league_position  \
0  Manchester City                Champions League                        1   
1        Liverpool                Champions League                        2   
2          Chelsea  Champions League Qualification                        3   
3        Tottenham  Champions League Qualification                        4   
4          Arsenal                   Europa League                        5   

   finance _live_games_televised  finance _tv_revenue  general_matches_played  \
0                             26            150986355                      38   
1                             29            152425146                      38   
2                             25            146030216                      38   
3                             26            145230801                      38   
4                             25            142193180                      38   

   general_won  general_draw  general_lost  attack_scored  ...  defence_saves  \
0           32             2             4             95  ...             58   
1           30             7             1             89  ...             76   
2           21             9             8             63  ...             85   
3           23             2            13             67  ...            121   
4           21             7            10             73  ...            133   

   defence_blocks  defence_interceptions  defence_tackles  \
0              63                    362              518   
1              78                    314              610   
2              73                    351              618   
3             112                    326              626   
4             118                    412              609   

   defence_tackles_last_man  defence_clearances  defence_clearances_headed  \
0                         0                 543                        295   
1                         2                 639                        317   
2                         2                 593                        330   
3                         3                 770                        411   
4                         2                 762                        404   

   defence_penalty_conceeded attack_posession  attack_pass_accuracy  
0                          4             64.0                  89.0  
1                          1             58.8                  84.4  
2                          2             59.9                  87.6  
3                          5             56.7                  83.2  
4                          7             56.1                  83.3  

[5 rows x 44 columns]

In [6]:
transfer_df.head()

team  end_2019  end_2018  end_2017  end_2016  end_2015  \
0    Manchester City    -22.99   -226.15   -178.15   -140.86    -57.68   
1  Manchester United    -52.15   -152.90   -137.85    -53.93   -146.06   
2          Tottenham      5.35    -17.70    -31.30     16.58     -4.32   
3          Liverpool   -143.70     20.62      5.48    -34.85    -52.16   
4            Chelsea   -148.25    -60.00    -24.40     -3.01      7.15   

   end_2014  end_2013  end_2012  end_2011  end_2010    Total  
0   -104.70    -17.65    -59.85   -146.46   -116.35 -1070.83  
1    -75.33    -65.57    -47.84    -11.99     77.07  -666.55  
2     13.55     -0.47     33.80    -23.67     -8.60   -16.78  
3    -25.50    -60.15    -43.63      3.98      5.05  -324.86  
4    -52.92    -84.25    -64.05   -105.00    -26.20  -560.93

### Retirando a virgula de algumas colunas que foram reconhecidas como objeto

In [7]:
columns = ["attack_passes", "attack_passes_long", "attack_passes_back", "defence_clearances"]
for column in columns:
    epl_df[column] = epl_df[column].str.replace(",", "").astype(int)

### Ambas as tabelas tem o nome do Time, precisamos tratar isso antes de unir para não duplicar dados

- o .strip() retira os espaços no começo e fim da string
- na tabela transfer_df o header Team está com letra minuscula

In [8]:
transfer_df.rename(columns={'team':'Team'}, inplace=True)


In [9]:
transfer_df["Team"] = transfer_df["Team"].str.strip()
transfer_df["Team"]

0       Manchester City
1     Manchester United
2             Tottenham
3             Liverpool
4               Chelsea
5               Arsenal
6               Burnley
7               Everton
8             Leicester
9      Newcastle United
10       Crystal Palace
11          Bournemouth
12             West Ham
13              Watford
14             Brighton
15         Huddersfield
16          Southampton
17        Wolverhampton
18              Cardiff
19               Fulham
Name: Team, dtype: object

In [10]:
epl_df["Team"] = epl_df.sort_values("Team")["Team"].str.strip()
epl_df

Team                        category  \
0     Manchester City                Champions League   
1           Liverpool                Champions League   
2             Chelsea  Champions League Qualification   
3           Tottenham  Champions League Qualification   
4             Arsenal                   Europa League   
5   Manchester United                   Europa League   
6       Wolverhampton     Europa League Qualification   
7             Everton            No UEFA Competitions   
8           Leicester            No UEFA Competitions   
9            West Ham            No UEFA Competitions   
10            Watford            No UEFA Competitions   
11     Crystal Palace            No UEFA Competitions   
12          Newcastle            No UEFA Competitions   
13        Bournemouth            No UEFA Competitions   
14            Burnley            No UEFA Competitions   
15        Southampton            No UEFA Competitions   
16           Brighton            No UEFA Competitions   
17            Cardiff                       Relegated   
18             Fulham                       Relegated   
19       Huddersfield                       Relegated   

    general_league_position  finance _live_games_televised  \
0                         1                             26   
1                         2                             29   
2                         3                             25   
3                         4                             26   
4                         5                             25   
5                         6                             27   
6                         7                             15   
7                         8                             18   
8                         9                             15   
9                        10                             16   
10                       11                             10   
11                       12                             12   
12                       13                             19   
13                       14                             10   
14                       15                             11   
15                       16                             10   
16                       17                             13   
17                       18                             12   
18                       19                             13   
19                       20                             10   

    finance _tv_revenue  general_matches_played  general_won  general_draw  \
0             150986355                      38           32             2   
1             152425146                      38           30             7   
2             146030216                      38           21             9   
3             145230801                      38           23             2   
4             142193180                      38           21             7   
5             142512868                      38           19             9   
6             127165114                      38           16             9   
7             128603905                      38           15             9   
8             123328078                      38           15             7   
9             122528663                      38           15             7   
10            113895527                      38           14             8   
11            114215215                      38           14             7   
12            120130418                      38           12             9   
13            108139973                      38           13             6   
14            107340558                      38           11             7   
15            104302937                      38            9            12   
16            105741728                      38            9             9   
17            102704107                      38           10             4   
18            101904692           

### Unindo as duas tabelas

In [11]:
result = pd.merge(epl_df, transfer_df, on="Team")
result

Team                        category  \
0     Manchester City                Champions League   
1           Liverpool                Champions League   
2             Chelsea  Champions League Qualification   
3           Tottenham  Champions League Qualification   
4             Arsenal                   Europa League   
5   Manchester United                   Europa League   
6       Wolverhampton     Europa League Qualification   
7             Everton            No UEFA Competitions   
8           Leicester            No UEFA Competitions   
9            West Ham            No UEFA Competitions   
10            Watford            No UEFA Competitions   
11     Crystal Palace            No UEFA Competitions   
12        Bournemouth            No UEFA Competitions   
13            Burnley            No UEFA Competitions   
14        Southampton            No UEFA Competitions   
15           Brighton            No UEFA Competitions   
16            Cardiff                       Relegated   
17             Fulham                       Relegated   
18       Huddersfield                       Relegated   

    general_league_position  finance _live_games_televised  \
0                         1                             26   
1                         2                             29   
2                         3                             25   
3                         4                             26   
4                         5                             25   
5                         6                             27   
6                         7                             15   
7                         8                             18   
8                         9                             15   
9                        10                             16   
10                       11                             10   
11                       12                             12   
12                       14                             10   
13                       15                             11   
14                       16                             10   
15                       17                             13   
16                       18                             12   
17                       19                             13   
18                       20                             10   

    finance _tv_revenue  general_matches_played  general_won  general_draw  \
0             150986355                      38           32             2   
1             152425146                      38           30             7   
2             146030216                      38           21             9   
3             145230801                      38           23             2   
4             142193180                      38           21             7   
5             142512868                      38           19             9   
6             127165114                      38           16             9   
7             128603905                      38           15             9   
8             123328078                      38           15             7   
9             122528663                      38           15             7   
10            113895527                      38           14             8   
11            114215215                      38           14             7   
12            108139973                      38           13             6   
13            107340558                      38           11             7   
14            104302937                      38            9            12   
15            105741728                      38            9             9   
16            102704107                      38           10             4   
17            101904692                      38            7             5   
18             96628865                      38            3             7   

    general_lost  attack_scored  ...  end_2018  end_2017  end_2016  end_20

### Tratando a categoria das ligas, que é objeto

In [12]:
category_encoding = pd.get_dummies(result["category"])
category_encoding

Champions League  Champions League Qualification  Europa League  \
0                  1                               0              0   
1                  1                               0              0   
2                  0                               1              0   
3                  0                               1              0   
4                  0                               0              1   
5                  0                               0              1   
6                  0                               0              0   
7                  0                               0              0   
8                  0                               0              0   
9                  0                               0              0   
10                 0                               0              0   
11                 0                               0              0   
12                 0                               0              0   
13                 0                               0              0   
14                 0                               0              0   
15                 0                               0              0   
16                 0                               0              0   
17                 0                               0              0   
18                 0                               0              0   

    Europa League Qualification  No UEFA Competitions  Relegated  
0                             0                     0          0  
1                             0                     0          0  
2                             0                     0          0  
3                             0                     0          0  
4                             0                     0          0  
5                             0                     0          0  
6                             1                     0          0  
7                             0                     1          0  
8                             0                     1          0  
9                             0                     1          0  
10                            0                     1          0  
11                            0                     1          0  
12                            0                     1          0  
13                            0                     1          0  
14                            0                     1          0  
15                            0                     1          0  
16                            0                     0          1  
17                            0                     0          1  
18                            0                     0          1

### Unindo os dados do Encoding

In [13]:
result_encoding = pd.concat([result, category_encoding], axis=1)

result_encoding = result_encoding.drop(["category", "Team"], axis = 1)
result_encoding

general_league_position  finance _live_games_televised  \
0                         1                             26   
1                         2                             29   
2                         3                             25   
3                         4                             26   
4                         5                             25   
5                         6                             27   
6                         7                             15   
7                         8                             18   
8                         9                             15   
9                        10                             16   
10                       11                             10   
11                       12                             12   
12                       14                             10   
13                       15                             11   
14                       16                             10   
15                       17                             13   
16                       18                             12   
17                       19                             13   
18                       20                             10   

    finance _tv_revenue  general_matches_played  general_won  general_draw  \
0             150986355                      38           32             2   
1             152425146                      38           30             7   
2             146030216                      38           21             9   
3             145230801                      38           23             2   
4             142193180                      38           21             7   
5             142512868                      38           19             9   
6             127165114                      38           16             9   
7             128603905                      38           15             9   
8             123328078                      38           15             7   
9             122528663                      38           15             7   
10            113895527                      38           14             8   
11            114215215                      38           14             7   
12            108139973                      38           13             6   
13            107340558                      38           11             7   
14            104302937                      38            9            12   
15            105741728                      38            9             9   
16            102704107                      38           10             4   
17            101904692                      38            7             5   
18             96628865                      38            3             7   

    general_lost  attack_scored  defence_goals_conceeded  \
0              4             95                       23   
1              1             89                       22   
2              8             63                       39   
3             13             67                       39   
4             10             73                       51   
5             10             65                       54   
6             13             47                       46   
7             14             54                       46   
8             16             51                       48   
9             16             52                       55   
10            16             52                       59   
11            17             51                       53   
12            19             56                       70   
13            20             45                       68   
14            17             45                       65   
15            20             35                       60   
16            24             34                       69   
17            26             34                       81   
18            28             22                       76   

 

## K-Means k=2

In [14]:
kmeans = KMeans(n_clusters=2, random_state=0).fit(result_encoding)

In [15]:
a = np.array(kmeans.labels_)
a.sum()

6

In [16]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 13, 1: 6}


Classe  Nº Instancias
0       0             13
1       1              6

### Silhouette

In [17]:
km_ = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
sk.metrics.silhouette_score(km_, labels=km_["cluster_number"])

0.7692706157307959

## K-Means k=5

In [18]:
kmeans = KMeans(n_clusters=5, random_state=0).fit(result_encoding)

In [19]:
a = np.array(kmeans.labels_)
a.sum()

37

In [20]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 7, 1: 1, 2: 1, 3: 6, 4: 4}


Classe  Nº Instancias
0       0              7
1       1              1
2       2              1
3       3              6
4       4              4

### Silhouette

In [21]:
km_ = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
sk.metrics.silhouette_score(km_, labels=km_["cluster_number"])

0.47013832606718325

## K-Means k=10

In [22]:
kmeans = KMeans(n_clusters=10, random_state=0).fit(result_encoding)

In [23]:
a = np.array(kmeans.labels_)
a.sum()

85

In [24]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 4, 1: 1, 2: 1, 3: 1, 4: 1, 5: 2, 6: 2, 7: 4, 8: 2, 9: 1}


Classe  Nº Instancias
0       0              4
1       1              1
2       2              1
3       3              1
4       4              1
5       5              2
6       6              2
7       7              4
8       8              2
9       9              1

### Silhouette

In [25]:
km_ = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
sk.metrics.silhouette_score(km_, labels=km_["cluster_number"])

0.46420915413668373

## K-Means k=19

In [26]:
kmeans = KMeans(n_clusters=19, random_state=0).fit(result_encoding)

In [27]:
a = np.array(kmeans.labels_)
a.sum()

171

In [28]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 1, 1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1, 11: 1, 12: 1, 13: 1, 14: 1, 15: 1, 16: 1, 17: 1, 18: 1}


Classe  Nº Instancias
0        0              1
1        1              1
2        2              1
3        3              1
4        4              1
5        5              1
6        6              1
7        7              1
8        8              1
9        9              1
10      10              1
11      11              1
12      12              1
13      13              1
14      14              1
15      15              1
16      16              1
17      17              1
18      18              1

In [29]:
# h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
# sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

## Kmeans k=7 e max_iter=1

In [30]:
kmeans = KMeans(n_clusters=7, max_iter=1).fit(result_encoding)

km_clustered = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
#Silhouette
sk.metrics.silhouette_score(km_clustered, labels=km_clustered["cluster_number"])

0.4555277338959717

In [31]:
a = np.array(kmeans.labels_)
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 6, 1: 2, 2: 3, 3: 1, 4: 1, 5: 2, 6: 4}


Classe  Nº Instancias
0       0              6
1       1              2
2       2              3
3       3              1
4       4              1
5       5              2
6       6              4

## Kmeans k=7 e max_iter=10

In [32]:
kmeans = KMeans(n_clusters=7, max_iter=10).fit(result_encoding)

km_clustered = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
#Silhouette
sk.metrics.silhouette_score(km_clustered, labels=km_clustered["cluster_number"])

0.4555277338959705

In [33]:
a = np.array(kmeans.labels_)
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 1, 1: 4, 2: 2, 3: 6, 4: 1, 5: 2, 6: 3}


Classe  Nº Instancias
0       0              1
1       1              4
2       2              2
3       3              6
4       4              1
5       5              2
6       6              3

## Kmeans k=7 e max_iter=100

In [34]:
kmeans = KMeans(n_clusters=7, max_iter=100).fit(result_encoding)

km_clustered = result_encoding.assign(cluster_number=kmeans.predict(result_encoding))
#Silhouette
sk.metrics.silhouette_score(km_clustered, labels=km_clustered["cluster_number"])

0.45552773389597045

In [35]:
a = np.array(kmeans.labels_)
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 3, 1: 2, 2: 6, 3: 2, 4: 1, 5: 4, 6: 1}


Classe  Nº Instancias
0       0              3
1       1              2
2       2              6
3       3              2
4       4              1
5       5              4
6       6              1

## Hierarquico k=2 / linkage=ward

In [36]:
clustering = AgglomerativeClustering(n_clusters=2, affinity='euclidean', linkage="ward").fit(result_encoding)


In [37]:
a = np.array(clustering.labels_)
a

array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [38]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 6, 1: 13}


Classe  Nº Instancias
0       0              6
1       1             13

### Silhouette

In [39]:
h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

0.7692706157307959

## Hierarquico k=2 / linkage=complete

In [40]:
clustering = AgglomerativeClustering(n_clusters=2, affinity='euclidean', linkage="complete").fit(result_encoding)


In [41]:
a = np.array(clustering.labels_)
a

array([0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])

In [42]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 6, 1: 13}


Classe  Nº Instancias
0       0              6
1       1             13

### Silhouette

In [43]:
h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

0.7692706157307959

## Hierarquico k=5 / linkage=complete

In [44]:
clustering = AgglomerativeClustering(n_clusters=5, affinity='euclidean', linkage="complete").fit(result_encoding)


In [45]:
a = np.array(clustering.labels_)
a

array([2, 1, 1, 1, 1, 3, 0, 0, 0, 0, 0, 0, 4, 4, 4, 4, 4, 4, 4])

In [46]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 6, 1: 4, 2: 1, 3: 1, 4: 7}


Classe  Nº Instancias
0       0              6
1       1              4
2       2              1
3       3              1
4       4              7

### Silhouette

In [47]:
h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

0.47013832606718325

## Hierarquico k=5 / linkage=ward

In [48]:
clustering = AgglomerativeClustering(n_clusters=5, affinity='euclidean', linkage="ward").fit(result_encoding)


In [49]:
a = np.array(clustering.labels_)
a

array([2, 1, 1, 1, 1, 4, 3, 3, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [50]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 10, 1: 4, 2: 1, 3: 3, 4: 1}


Classe  Nº Instancias
0       0             10
1       1              4
2       2              1
3       3              3
4       4              1

### Silhouette

In [51]:
h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

0.5012077665339084

## Hierarquico k=19 / linkage=single

In [52]:
clustering = AgglomerativeClustering(n_clusters=19, affinity='euclidean', linkage="single").fit(result_encoding)


In [53]:
a = np.array(clustering.labels_)
a

array([11, 13, 17,  9, 15, 16, 10, 14,  4,  7, 18,  8, 12,  6,  3,  1,  5,
        2,  0])

In [54]:
unique, counts = np.unique(a, return_counts=True)
dicionario = dict(zip(unique, counts))
print(dicionario)
lista = list(dicionario.items())
class_instances = pd.DataFrame(lista, columns=["Classe", "Nº Instancias"])
class_instances

{0: 1, 1: 1, 2: 1, 3: 1, 4: 1, 5: 1, 6: 1, 7: 1, 8: 1, 9: 1, 10: 1, 11: 1, 12: 1, 13: 1, 14: 1, 15: 1, 16: 1, 17: 1, 18: 1}


Classe  Nº Instancias
0        0              1
1        1              1
2        2              1
3        3              1
4        4              1
5        5              1
6        6              1
7        7              1
8        8              1
9        9              1
10      10              1
11      11              1
12      12              1
13      13              1
14      14              1
15      15              1
16      16              1
17      17              1
18      18              1

### Silhouette

In [55]:
# h_clustered = result_encoding.assign(cluster_number=clustering.fit_predict(result_encoding))
# sk.metrics.silhouette_score(h_clustered, labels=h_clustered["cluster_number"])

## Considerando os dados acima

Podemos observar que os melhores resultados foram:

- Kmeans = 0.7692706157307959
- Hierarquico = 0.7692706157307959

E que os resultados com as mudanças no linkage tiveram seu pior resultado com single

As alterações em max_iter foram mais signficativas com k=5 ou k=7. Quando k=2 os resultados foram identicos.